### PR TITLE
[Enhancement] runtime training sample generation for BERT

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
This PR tokenizes and maps tokens to integers before creating masks in the trainining samples. It makes the pre-training data generation 50% faster. It also add an option to runtime training data generation, so that we can pre-train even larger corpus.

change list:
- remove recordio format output, as it is the most space consuming option
- perform vocabulary look up first, before selecting next sentences/masks. Manipulating list of ints instead of string is much more efficient. 
- added `--raw` option to the pretraining script to allow runtime training sample generation.

future work:
- we can add `--int-inputs` option to allow data source preprocessed (tokenized and replaced with word ids) and stored as integers. Currently vocabulary lookup still accounts for significant amount of processing time. 

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by mli** [comment]
Job PR-687/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-687/5/index.html

**Comment by mli** [comment]
Job PR-687/1 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-687/1/index.html

**Comment by mli** [comment]
Job PR-687/2 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-687/2/index.html

**Comment by eric-haibin-lin** [comment]
@araitats you might be interested in this PR

File: scripts/bert/fp16_utils.py [hunk 1]
```diff
@@ -245,8 +245,9 @@ def update_scale(self, overflow):
                 self.loss_scale /= self.scale_factor
                 self._last_rescale_iter = self._num_steps
                 self._overflows_since_rescale = 0
-                logging.info('DynamicLossScaler: overflow detected. set loss_scale = %s',
-                             self.loss_scale)
+                if self.loss_scale < 1:
+                    logging.info('DynamicLossScaler: overflow detected. set loss_scale = %s',
```
Comment (bikestra) [inline scripts/bert/fp16_utils.py]: Should we rather `warn`?

File: scripts/bert/run_pretraining_hvd.py [hunk 2]
```diff
@@ -43,9 +44,23 @@
...
+from pretraining_utils import get_online_pretrain_dataset
 
 # parser
 parser = get_argparser()
+parser.add_argument('--raw', action='store_true',
+                    help='Input raw text files instead of pre-processed npz files')
+parser.add_argument('--max_seq_length', type=int, default=None,
+                    help='Maximum input sequence length.')
+parser.add_argument('--short_seq_prob', type=float, default=None,
+                    help='Probability for short sequences.')
```
Comment (bikestra) [inline scripts/bert/run_pretraining_hvd.py]: The same comment about the description of this parameter.

**Comment by mli** [comment]
Job PR-687/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-687/5/index.html

**Comment by mli** [comment]
Job PR-687/6 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-687/6/index.html

File: scripts/bert/create_pretraining_data.py [hunk 3]
```diff
@@ -331,45 +202,48 @@ def create_training_instances(x):
...
+
 
     # Account for [CLS], [SEP], [SEP]
     max_num_tokens = max_seq_length - 3
 
     # We *usually* want to fill up the entire sequence since we are padding
     # to `max_seq_length` anyways, so short sequences are generally wasted
     # computation. However, we *sometimes*
-    # (i.e., short_seq_prob == 0.1 == 10% of the time) want to use shorter
+    # (i.e., short_seq_prob == 0.1, 10% of the time) want to use shorter
```
[the remark this answers is not in the document]
Comment (eric-haibin-lin) [reply scripts/bert/create_pretraining_data.py]: It was not mentioned in the paper, but found in author's implementation :)

**Comment by mli** [comment]
Job PR-687/7 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-687/7/index.html

**Comment by mli** [comment]
Job PR-687/8 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-687/8/index.html

**Comment by mli** [comment]
Job PR-687/9 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-687/9/index.html

**Comment by mli** [comment]
Job PR-687/10 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-687/10/index.html

**Comment by mli** [comment]
Job PR-687/11 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-687/11/index.html

**Comment by mli** [comment]
Job PR-687/12 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-687/12/index.html

**Comment by mli** [comment]
Job PR-687/13 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-687/13/index.html

**Comment by mli** [comment]
Job PR-687/14 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-687/14/index.html

**Comment by mli** [comment]
Job PR-687/15 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-687/15/index.html

**Comment by mli** [comment]
Job PR-687/16 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-687/16/index.html

**Comment by haven-jeon** [comment]
Looks great.

**Comment by mli** [comment]
Job PR-687/17 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-687/17/index.html
